File: 933-increasing-order-search-tree/increasing-order-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.*;

class Solution {
    public TreeNode increasingBST(TreeNode root) {

        Stack<TreeNode> stack = new Stack<>();

        TreeNode current = root;

        TreeNode dummy = new TreeNode(0);
        TreeNode tail = dummy;

        while (current != null || !stack.isEmpty()) {

            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            current = stack.pop();

            TreeNode right = current.right;

            current.left = null;

            tail.right = current;
            tail = current;

            current = right;
        }

        return dummy.right;
    }
}

File: 933-increasing-order-search-tree/increasing-order-search-tree.py
class Solution:
    def increasingBST(self, root):
        stack = []
        curr = root

        dummy = TreeNode(0)
        tail = dummy

        while curr or stack:

            while curr:
                stack.append(curr)
                curr = curr.left

            curr = stack.pop()

            right = curr.right

            curr.left = None

            tail.right = curr
            tail = curr

            curr = right

        return dummy.right